### PR TITLE
Fixes #506 - corrects compilation of multi-assignments in which one LHS is a subexpression of another LHS

### LINKED ISSUE #506: Parallel assignment occurs sequentially #bug
Dafny assigns sequentially instead of what supposed to be a parallel assignment.

This is an example of a code that runs just fine with no RunTimeErrors.
https://www.riseforfun.com/Dafny/mieKE

And this is the same code with a change in the order of assignment in line 117. **This code has an RTE.**
https://www.riseforfun.com/Dafny/uaMz

### PR DESCRIPTION
**Pull request by davidcok**
Fixes issue #506, in which multi-assignments with non-independent left-hand-sides were compiled incorrectly. Static verification was correct. Tests added also.

### CODE REVIEW

File: Source/Dafny/Compiler.cs [hunk 1]
```diff
@@ -342,14 +350,25 @@ protected interface IClassWriter {
...
       return w;
     }
 
-    protected virtual void EmitMultiAssignment(List<ILvalue> lhss, List<Type> lhsTypes, out List<TargetWriter> wRhss, List<Type> rhsTypes, TargetWriter wr) {
+    protected virtual string EmitAssignmentLhs(Expression e, TargetWriter wr) {
+      string target = idGenerator.FreshId("_lhs");
+      wr.Write(GenerateLhsDecl(target, e.Type, wr, null));
+      wr.Write(" = ");
+      TrExpr(e,wr,false);
+      EndStmt(wr);
```
Comment (RustanLeino) [inline Source/Dafny/Compiler.cs:366]: Use `EmitAssignment` for these 4 lines (or, if for some reason that doesn't work, at least use `EmitAssignmentRhs` for the last 3 of these lines).

File: Source/Dafny/Compiler.cs [hunk 2]
```diff
@@ -363,9 +382,57 @@ protected interface IClassWriter {
...
+            lhssn.Add(newLhs);
+          } else if (lexpr is MultiSelectExpr) {
+            MultiSelectExpr seqExpr = lexpr as MultiSelectExpr;
+            Expression array = seqExpr.Array;
+            List<Expression> indices = seqExpr.Indices;
+            string targetArray = idGenerator.FreshId("_lhs");
+            wr.Write(GenerateLhsDecl(targetArray, array.Type, wr, null));
+            wr.Write(" = ");
+            TrExpr(array,wr,false);
+            EndStmt(wr);
```
Comment (RustanLeino) [inline Source/Dafny/Compiler.cs:416]: See comment on L363

File: Source/Dafny/Compiler-java.cs [hunk 3]
```diff
@@ -3757,5 +3757,13 @@ protected override BlockTargetWriter CreateIterator(IteratorDecl iter, TargetWri
     {
       throw new NotImplementedException();
     }
+
+    /// <summary>
+    /// Transforms a legal file name (without extension or directcory) into
+    /// a legal class name in the target language
+    /// </summary>
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Compiler-java.cs:3764]: OK

File: Source/Dafny/Compiler.cs [hunk 4]
```diff
@@ -102,6 +102,14 @@ protected interface IClassWriter {
     protected IClassWriter CreateClass(string moduleName, string name, List<TypeParameter>/*?*/ typeParameters, TargetWriter wr) {
       return CreateClass(moduleName, name, false, null, typeParameters, null, null, wr);
     }
+    
+        
+    // Transforms a legal file name (without extension or directcory) into 
+    // a legal class name in the target language
+    public virtual string TransformToClassName(string baseName) {
+      return baseName;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Compiler.cs:110]: OK

File: Test/git-issues/git-issue-506.dfy [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+method Main() {
+
+  var a := new int[10];
+  var index := 6;
+  a[8] := 1;
+  a[index], index := 3, index+1;
+  assert a[6] == 3;
+  assert index == 7;
+  print index, " ", a[6], a[7], a[8], "\n";  // Should be: "7 301"
+  index, a[index] := index+1, 9;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-506.dfy:16]: OK

File: .gitignore [hunk 6]
```diff
@@ -40,6 +40,7 @@ Test/*/bin/
 Test/*/obj/
 Test/*/obj_core/
 Test/packages
+Test/traits/TraitCompile.cs
```
[the remark this answers is not in the document]
Comment (davidcok) [reply .gitignore:43]: Agreed - a separate task to clean up the .gitignore would be good, but not here

File: Source/Dafny/Compiler.cs [hunk 7]
```diff
@@ -342,14 +350,25 @@ protected interface IClassWriter {
...
     protected virtual TargetWriter EmitAssignmentRhs(TargetWriter wr) {
       wr.Write(" = ");
       var w = wr.Fork();
       EndStmt(wr);
       return w;
     }
 
-    protected virtual void EmitMultiAssignment(List<ILvalue> lhss, List<Type> lhsTypes, out List<TargetWriter> wRhss, List<Type> rhsTypes, TargetWriter wr) {
+    protected virtual string EmitAssignmentLhs(Expression e, TargetWriter wr) {
+      string target = idGenerator.FreshId("_lhs");
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Compiler.cs:362]: OK, though I generally prefer not to have to do the inference myself, so I often like explicit types.

File: Source/Dafny/Compiler.cs (hunk 2 above)
Comment (davidcok) [reply Source/Dafny/Compiler.cs:416]: Changed to use EmitAssignmentLhs

File: Source/Dafny/Compiler.cs (hunk 1 above)
Comment (davidcok) [reply Source/Dafny/Compiler.cs:366]: There is no signature for EmitAssignment or EmitAssignmentRhs that works here

File: Source/Dafny/Compiler.cs [hunk 8]
```diff
@@ -363,9 +382,57 @@ protected interface IClassWriter {
...
+      if (lhss.Count > 1) {
+        lhssn = new List<ILvalue>();
+        for (int i = 0; i < lhss.Count; ++i) {
+          Expression lexpr = lhsExprs[i];
+          ILvalue lhs = lhss[i];
+          if (lexpr is NameSegment || lexpr is IdentifierExpr) {
+            lhssn.Add(lhs);
+          } else if (lexpr is MemberSelectExpr) {
+            lhssn.Add(lhs);
+          } else if (lexpr is ExprDotName) {
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/Dafny/Compiler.cs:395]: Done. I had thought to get a tutorial from you on ASTs yesterday, but we ran out of time.

File: Source/Dafny/Compiler.cs (hunk 1 above)
Comment (RustanLeino) [reply Source/Dafny/Compiler.cs:368]: Does `EmitAssignmentRhs(e, false, wr);` do what L366-L368 do for all languages?

File: Test/git-issues/git-issue-506.dfy [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  assert o.f == 6;
+  assert ooo.f == 7;
+  expect ooo.f == 7;
+  print o.f, " ", ooo.f, "\n";
+
+  var aa := new int[9,9];
+  var j := 4;
+  var k := 5;
+  aa[j,k] := 8;
+  j, k, aa[j,k] := 2, 3, 7;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/git-issue-506.dfy:48]: Various additional tests added

**Pull request merged**